### DISCUSSION TITLE
Prototype 기반 상속에 대해 설명해주세요.

### DISCUSSION THREAD

**Discussion by BaekYuri** [wanted-9th-3team/frontend-study#34] [프론트 기술 질문 공간]
Prototype 기반 상속에 대해 설명해주세요.

**Comment by michoball** [comment]
어떤 객체에 대해 확장된 새로운 객체를 생성하여 객체 내에 정의된 변수, 메서드 함수등를 상속받아 쓰고자 할 때 사용합니다. 
javascript가 프로토타입 기반 언어이기 때문에 es6 이후 클래스 문법이 나오기 전에는 프로토타입 기반으로 객체를 상속받아 새로운 객체가 상속 받아온 객체의 기능을 사용할 수 있도록 하였습니다. 

객체의 프로토타입 객체에 접근하기 위해서는 객체.__proto__ 를 사용합니다. es6이후에는  Object.getPrototypeOf(obj) 를 통해 접근이 가능합니다. 

어떤객체.prototype.뭐시기 로 새로 무언가를 정의하면 어떤객체의 prototype안에 새로운 변수나 함수가 생겨서 이 객체를 상속받는 모든 객체는 이 변수나 함수를 사용할 수 있습니다. 

음... 어렵네요!

**Comment by javascriptbiggosoo** [comment]
자바스크립트의 _reference_ 타입은 _object_ 타입이라고도 하는데요.
자바스크립트의 모든 _reference_ 타입은 내장 `Object` constructor 함수에서 생성되며 다시 그 하위에서 `Array`, `Data`, `Function ` 등으로 구분될 수 있습니다.
(해당 _object_ 들은 프로퍼티와 메소드를 갖고 있습니다.)
리터럴로 생성되는 경우에도 자동으로 이런 내장 _object_ 들을 상속받아 거기에 포함된(우리가 작성하지 않은) 메소드를 사용할 수 있게 됩니다.
object에서 메소드를 호출하는 경우 먼저 그 object에 해당 메소드가 있는지를 확인한 후, 없다면 상위 프로토타입에 메소드가 있는지 확인한 후 있다면 실행, 없다면 다시 상위를 찾아가는 과정을 반복합니다.

ES6에서 출시된 `class` 는 사실 BTS(behind the scene)에서 바로 이 constructor 함수, prototype이 작동하는 것이기 때문에 자바스크립트는 _class_ 가 없는 언어라고 놀리는 개발자들이 있습니다.

**Comment by skyhanull** [comment]
object에서 프로퍼티를 읽으려고 하는데 해당 프로퍼티가 없으면 자바스크립트는 자동으로 프로토타입에서 프로퍼티를 찾기 때문이죠. 프로그래밍에선 이런 동작 방식을 '프로토타입 상속’이라 부릅니다.
객체의 프로토타입에 접근하기 위해서 _proto_라는 속성을 사용하여 접근할 수 있습니다

**Comment by bona373737** [comment]
자바같은 객체지향언어에서 클래스로 객체를 정의하고 상속받아서 재사용,확장시켜 사용하는 프로그래밍 방식이 자바스크립트에서는 프로토타입이라는 속성으로 구현된다.

자바스크립트의 모든 객체엔 숨겨진 속성값으로 프로토타입을 갖는다. 기본값은 null이고, "객체._proto_ = 상속받으려는 객체" 의 방식으로 프로토타입을 설정하여 다른 객체의 변수와 메서드를 상속받아 사용할 수 있다. 

ES6에서 클래스라는 개념이 도입되었는데 그 내부적으로는 여전히 프로토타입기반으로 구현된것으로 알고 있습니다.
